### PR TITLE
Inter-option dependency system

### LINKED ISSUE #74: Support for inter-option dependencies in value parsers
When building CLI tools that mirror Git's interface, it's common to have a global option like `-C <path>` that changes the working directory for subsequent operations. Ideally, value parsers (like `gitBranch()`) should be able to reference the parsed value of `-C` to validate and provide completions from the correct repository.

```shell
myapp -C ./path/to/repo branch --delete [completions from ./path/to/repo]
```

Currently, Optique binds configuration (like `dir`) at parser *construction* time, making it impossible for one option's parser to depend on another option's parsed value.

## Use cases

- Git-like CLIs where `-C <path>` changes the repository context
- Database CLIs where `--database <name>` affects table/column completions
- Cloud CLIs where `--region <region>` affects resource name validation

### 1. Dynamic directory resolver (function-based)

Allow `dir` to accept a function instead of a static string:

```typescript
gitBranch({ dir: () => resolveFromArgs() })
```

*Rejected:* The function has no access to the parsing context, so it would need to manually parse `process.argv`, duplicating work and risking inconsistency with the actual parser state.

### 2. Parsing context injection

Extend the value parser interface to receive a parsing context containing raw arguments:

```typescript
interface ParsingContext {
  readonly buffer: readonly string[];
  readonly originalArgs: readonly string[];
}
```

*Rejected:* Value parsers would need to re-implement option parsing logic to extract values like `-C`. This duplicates parser functionality and is error-prone (handling `--key=value` vs `--key value`, short option combining, etc.).

### 3. Post-parsing validation

Parse options separately, then validate in application code.

*Rejected:* Loses the benefits of parse-time validation and, more critically, shell completion.

### 4. Completion-time context awareness

Pass raw arguments to the `suggest()` function only.

*Rejected:* Only solves completions, not parse-time validation. Inconsistent behavior between validation and completion is confusing.

## Summary

This is a genuinely difficult problem that touches on fundamental aspects of Optique's architecture. All approaches considered so far either duplicate parsing logic, break the abstraction, or solve only part of the problem.

Further brainstorming is needed to find a solution that:

- Provides access to other options' parsed values (not raw strings)
- Works for both validation and completion
- Doesn't require value parsers to understand option syntax
- Fits naturally into Optique's combinatorial design

## References

- Original feedback: https://github.com/dahlia/optique/issues/71#issuecomment-3708538910

### PR DESCRIPTION
**Pull request by dahlia**
## Summary

This PR adds inter-option dependencies for value parsers, enabling one option's valid values to depend on another option's parsed value. This addresses a fundamental limitation where configuration was bound at parser construction time.

Closes #74

## Key features

- `dependency(parser)`: Wraps a value parser to create a dependency source that can be referenced by derived parsers
- `source.derive(options)`: Creates a derived parser whose valid values depend on the source's value
- `deriveFrom(options)`: Creates a parser that depends on multiple dependency sources
- Deferred parsing: Dependent options store raw input during parsing, then re-validate after all options are collected
- Order independence: Options can appear in any order on the command line
- Context-aware shell completion: Suggestions reflect already-provided dependency values

## Example

```typescript
import { dependency } from "@optique/core/dependency";
import { object, option, choice } from "@optique/core";

const modeParser = dependency(choice(["dev", "prod"] as const));

const logLevelParser = modeParser.derive({
  metavar: "LEVEL",
  factory: (mode) => choice(
    mode === "dev" 
      ? ["debug", "info", "warn", "error"]
      : ["warn", "error"]
  ),
  defaultValue: () => "dev" as const,
});

const parser = object({
  mode: option("--mode", modeParser),
  logLevel: option("--log-level", logLevelParser),
});

// --mode dev --log-level debug ✓
// --mode prod --log-level debug ✗ (invalid in prod mode)
```

## Use cases addressed

- Git-like CLIs where `-C <dir>` changes the repository context for branch/tag completion
- Database CLIs where `--database` affects table/column completions  
- Cloud CLIs where `--region` affects resource name validation
- Build systems where `--target` determines valid architectures and optimization levels

### New module: `@optique/core/dependency`
- `dependency()` – Create a dependency source
- `derive()`/`deriveSync()`/`deriveAsync()` – Single-source derived parsers
- `deriveFrom()`/`deriveFromSync()`/`deriveFromAsync()` – Multi-source derived parsers

### Modified modules
- *constructs.ts* – Dependency resolution in `object()`, support across `merge()`, `concat()`, `or()`
- *primitives.ts* – Handle dependency sources in `option()` and `argument()`
- *modifiers.ts* – Propagate dependency state through `optional()`, `withDefault()`, `multiple()`, `map()`

### Documentation
- New concept guide: *docs/concepts/dependencies.md*
- Cookbook section: *Inter-option value dependencies*
- Example: *examples/patterns/inter-option-dependencies.ts*

### Tests
- ~12,000 lines of comprehensive test coverage
- Edge cases: diamond dependencies, circular prevention, async factories, modifier chains
- Integration: shell completion, help generation, error messages

## Limitations

- No nested dependencies (derived parsers cannot be dependency sources)
- `deriveFrom()` requires `DependencySource` objects, not derived parsers

## Breaking changes

None. This is a new additive feature.

### CODE REVIEW

File: packages/core/src/dependency.test.ts [hunk 1]
```diff
@@ -0,0 +1,2977 @@
...
+    if (result.success) {
+      assert.equal(result.value.mode, "prod");
+      assert.equal(result.value.logLevel, "quiet");
+      assert.equal(result.value.port, 443);
+      assert.equal(result.value.timeout, 1000);
+    }
+  });
+
+  // TODO: deriveFrom with multiple independent dependency sources not yet working correctly
+  test("deriveFrom with multiple independent dependency sources", async () => {
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/dependency.test.ts]: Addressed in d62ac4bfa95d74cd272ab7c60fbbdf1057c606ac.

File: packages/core/src/constructs.ts [hunk 2]
```diff
@@ -2521,41 +2936,229 @@ export function object<
...
+            isPendingDependencySourceState(originalFieldState[0]);
+          const wasPreCompletedCase2 = originalFieldState === undefined &&
+            isPendingDependencySourceState(fieldParser.initialState);
+          const wasPreCompletedCase3 = originalFieldState === undefined &&
+            isWrappedDependencySource(fieldParser);
+
+          if (
+            isDependencySourceState(fieldResolvedState) &&
+            (wasPreCompletedCase1 || wasPreCompletedCase2 ||
+              wasPreCompletedCase3)
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Addressed in aeb3e4c8d79fa1413b2efd902e47f875584f31aa.

File: docs/concepts/dependencies.md [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+const logLevelParser = modeParser.derive({
+  metavar: "LEVEL",
+  factory: (mode) => {
+    // Return different choices based on the mode value
+    if (mode === "dev") {
+      return choice(["debug", "info", "warn", "error"]);
+    } else {
+      return choice(["warn", "error"]);
+    }
+  },
```
[the remark this answers is not in the document]
Comment (dahlia) [reply docs/concepts/dependencies.md]: Addressed in 7b6d2b98a73901dc75d53c873b9d8fb6f187b50b.

File: examples/patterns/inter-option-dependencies.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+const logLevelParser = modeParser.derive({
+  metavar: "LEVEL",
+  factory: (mode) => {
+    // In dev mode, allow verbose logging options
+    if (mode === "dev") {
+      return choice(["debug", "info", "warn", "error"] as const);
+    }
+    // In prod mode, only allow less verbose options
+    return choice(["warn", "error"] as const);
+  },
```
[the remark this answers is not in the document]
Comment (dahlia) [reply examples/patterns/inter-option-dependencies.ts]: Addressed in 7b6d2b98a73901dc75d53c873b9d8fb6f187b50b.

File: packages/core/src/constructs.ts [hunk 5]
```diff
@@ -2101,6 +2221,301 @@ async function* suggestObjectAsync<
...
+    const resolved: Record<string | symbol, unknown> = {};
+    for (const key of Reflect.ownKeys(state)) {
+      resolved[key] = resolveDeferred(state[key], registry);
+    }
+    return resolved;
+  }
+
+  // Everything else (primitives, class instances) - return as-is
+  return state;
+}
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts:2387]: Addressed in c474bde05ceba32f7dce9df9b91f3b8b2a245833.

File: packages/core/src/constructs.ts [hunk 6]
```diff
@@ -2101,6 +2221,301 @@ async function* suggestObjectAsync<
...
+    await Promise.all(
+      keys.map(async (key) => {
+        resolved[key] = await resolveDeferredAsync(state[key], registry);
+      }),
+    );
+    return resolved;
+  }
+
+  return state;
+}
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts:2466]: Addressed in f750e8e51b0697127e5cd704fe2ce18743759039.
